Objective-C/CBLCoreMLPredictiveModel.m

```diff
@@ -0,0 +1,538 @@
+//  limitations under the License.
+//
+
+#import "CBLCoreMLPredictiveModel.h"
+#import "CBLCoreMLPredictiveModel+Internal.h"
+#import "CBLBlob.h"
+#import "CBLDatabase+Prediction.h"
+#import "CBLDictionary.h"
+#import "CBLMutableArray.h"
+#import "CBLMutableDictionary.h"
+
+#import <CoreImage/CoreImage.h>
+#import <Vision/Vision.h>
+
+#define kCBLMayUseVisionModel YES
+#define kCBLMaxVisionPredictedProbabilities 10
+
+@implementation CBLCoreMLPredictiveModel {
+    MLModel* _model;
+    VNCoreMLModel* _vnModel;
+    NSString* _vnFeatureName;
+}
+
+- (instancetype) initWithMLModel: (MLModel*)model {
+    self = [super init];
+    if (self) {
+        _model = model;
+        [self setupVisionModel];
+    }
+    return self;
+}
+
+
+- (void) setupVisionModel {
+    if (!kCBLMayUseVisionModel)
+        return;
+        
+    MLModelDescription* desc = _model.modelDescription;
+    NSDictionary<NSString*, MLFeatureDescription*>* inputSpecs = desc.inputDescriptionsByName;
+    NSDictionary<NSString*, MLFeatureDescription*>* outputSpecs = desc.outputDescriptionsByName;
+    
+    // Use Vision Model when there is single input and the input type is image:
+    if (inputSpecs.count == 1 && inputSpecs.allValues[0].type == MLFeatureTypeImage) {
+        NSError* error;
+        _vnModel = [VNCoreMLModel modelForMLModel: _model error: &error];
+        if (!_vnModel) {
+            CBLWarn(Query, @"Cannot create vision model from the model: %@. "
+                            "The regular MLModel will be used instead.", error);
+            return;
+        }
+        
+        BOOL isSupported = (desc.predictedFeatureName || outputSpecs.count == 1);
+        if (isSupported) {
+            _vnFeatureName = inputSpecs.allKeys[0];
+        } else {
+            CBLWarn(Query, @"Vision models that returns multiple VNCoreMLFeatureValueObservation "
+                            "or VNPixelBufferObservation are not supported. The regular MLModel will "
+                            "be used instead.");
+            _vnModel = nil;
+        }
+    }
+}
+
+
+#pragma mark - CBLPredictiveModel
+
+
+- (nullable CBLDictionary*) predict: (CBLDictionary*)input {
+    if (_vnModel)
+        return [self predictUsingVNModel: input];
+    else
+        return [self predictUsingMLModel: input];
+}
+
+
+- (nullable CBLDictionary*) predictUsingMLModel: (CBLDictionary*)input {
+    id spec = _model.modelDescription.inputDescriptionsByName;
+    id<MLFeatureProvider> features = [self.class featuresFromDictionary: input spec: spec];
+    if (!features)
+        return nil;
+    
+    NSError* error;
+    id<MLFeatureProvider> output = [_model predictionFromFeatures: features error: &error];
+    if (!output)
+        return nil;
+    
+    return [self.class dictionaryFromFeatures: output];
+}
+
+
+- (nullable CBLDictionary*) predictUsingVNModel: (CBLDictionary*)input {
+    assert(_vnModel != nil && _vnFeatureName != nil);
```

Comment:
for the consistency, should we use ??
`CBLAssertNotNil(_vnModel);` 
`CBLAssertNotNil(_vnFeatureName);`

Reply:
No. It serves different purpose. Here is code assertion (not API non-null validation).